content/blog/2023/2023-09-16-create-openfeign-http-client.md

```diff
@@ -0,0 +1,503 @@
+---
+authors: [sagaofsilence]
+title: "Create an HTTP Client with OpenFeign"
+categories: ["Java"]
+date: 2023-09-16 00:00:00 +0530
+excerpt: "Feign is an HTTP client that allows us to easily make web requests. With its simplified and fluent API, it makes consuming RESTful services a breeze."
+image: images/stock/0134-client_sever.png
+url: create-openfeign-http-client
+---
+
+## Introduction
+Feign is an open-source Java library that simplifies the process of making web requests. It streamlines the implementation of RESTful web services by providing a higher-level abstraction. Feign eliminates the need for boilerplate code, which makes the codebase more readable and maintainable.
+
+## What is Feign?
+Feign is a popular Java HTTP client library that offers several advantages and features, making it a good choice for developers building HTTP-based microservices and applications.
+
+### What is a declarative HTTP client?
+It's a way to make http requests by writing a java interface. Feign generates the actual implementation based on annotations that we provide.
+
+### Why use Feign?
+If we have a large set of APIs to call, we don't want to generate the HTTP code by hand or with code generation. It would be much easier and more maintainable to describe the API in a simple, small interface and let Feign generate the code.
+
+### Who should use Feign?
+If we are making http requests in our Java code, and don't want to write boilerplate code, or use libraries like Apache httpclient directly, Feign is a great choice.
+
+<a name="example-code" />
+{{% github "https://github.com/thombergs/code-examples/tree/master/openfeign/openfeign-client-intro" %}}
+
+## Creating a Basic Feign Client
+### Step 1: Add Feign Dependency
+    Include Feign library in the Maven pom.xml file as a dependency.
+```xml
+<dependency>
+    <groupId>io.github.openfeign</groupId>
+    <artifactId>feign-core</artifactId>
+    <version>12.5</version>
+</dependency>
+```
+### Step 2: Define Service Interface
+Define service interface. **It typically contains the method declarations annotated with Feign annotations.**
+
+We are going to declare an interface that defines signature of methods we want to call. These are just declarations. We do not implement those methods. The REST service implements those methods. These methods reperesnt the endpoints we want to call. **The method signature should include the HTTP method as well as all required data.**
+
+Let us define an interface to represent calculator service. It has simple API methods to perfom calculations like add, substract, multiply and divide.
+
+We would use Wiremock to emulate the service implementation. [WireMock](https://wiremock.org/) is a web service mocking and stubbing tool. It works by emulating a real HTTP server to which the test code can connect as if it were a real online service. It allows for HTTP response stubbing, request verification, proxy/interception, stub recording/playback, and fault injection.
```

Comment:
Why introduce Wiremock here? It's a valuable tool, but what does it have to do in the section "Define Service Interface", especially since you're not mentioning it again in the rest of the article. If you want to mention it, you should dedicate a whole section "Testing a Feign Client" to it.

Reply:
Added section "Notes on testing" after the steps.